subsys/sd/mmc.c

```diff
@@ -34,14 +50,491 @@ inline int mmc_ioctl(struct sd_card *card, uint8_t cmd, void *buf)
+	csd->write_current_min = (uint8_t)((raw_csd[1U] &
+				0x00E00000U) >> 21U);
+	csd->write_current_max = (uint8_t)((raw_csd[1U] &
+			        0x001C0000U) >> 18U);
+	csd->dev_size_mul = (uint8_t)((raw_csd[1U] &
+				0x00038000U) >> 15U);
+	csd->erase_size = (uint8_t)((raw_csd[1U] &
+				0x00007C00U) >> 10U);
+	csd-> write_prtect_size = (uint8_t)(raw_csd[1U] &
+		       		0x0000001FU);
+	csd->write_speed_factor = (uint8_t)((raw_csd[0U] &
+			       	0x1C000000U) >> 26U);
+	csd->write_blk_len = (uint8_t)((raw_csd[0U] &
+			       	0x03C00000U) >> 22U);
+	csd->file_fmt = (uint8_t)((raw_csd[0U] &
+				0x00000C00U) >> 10U);
+}
+
+static inline int mmc_set_max_freq(
+		struct sd_card *card, struct sd_csd *card_csd)
+{
+	int ret;
+	enum mmc_csd_freq frequency_code = (card_csd->xfer_rate & 0x7);
+	enum mmc_csd_freq multiplier_code = (card_csd->xfer_rate & 0x78);
+
+	/* 4.3 - 5.1 emmc spec says 26 MHz */
+	if (frequency_code == MMC_MAXFREQ_10MHZ &&
+			multiplier_code == MMC_MAXFREQ_MULT_26)
+	{
+		card->bus_io.clock = 26000000;
+		card->bus_io.timing = SDHC_TIMING_LEGACY;
+	}
+	/* 4.0 - 4.2 emmc spec says 20 MHz */
+	else if (frequency_code == MMC_MAXFREQ_10MHZ &&
+                        multiplier_code == MMC_MAXFREQ_MULT_20)
+	{
+		card->bus_io.clock = 20000000;
+                card->bus_io.timing = SDHC_TIMING_LEGACY;
+	}
+	else
+	{
+		LOG_INF ("Using Legacy MMC will have slow initialization");
+		return 0;
+	}
+
+	ret = sdhc_set_io(card->sdhc, &card->bus_io);
+	if (ret) {
+		LOG_ERR("Error seetting initial clock frequency");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int mmc_set_timing(struct sd_card *card)
+{
+	int ret;
+	struct sdhc_command cmd = {0};
+
+	cmd.opcode = SD_SWITCH;
+	cmd.arg = MMC_SWITCH_HS_TIMING_ARG;
+	cmd.response_type = SD_RSP_TYPE_R1b;
+	cmd.timeout_ms = CONFIG_SD_CMD_TIMEOUT;
+	ret = sdhc_request(card->sdhc, &cmd, NULL);
+	if (ret) {
+		LOG_DBG("Error setting bus timing: %d", ret);
+		return ret;
+	}
+	sdmmc_wait_ready(card);
+	card->bus_io.clock = 52000000;
+        card->bus_io.timing = SDHC_TIMING_HS;
+	ret = sdhc_set_io(card->sdhc, &card->bus_io);
+        if (ret) {
+                return ret;
+        }
+
+	return 0;
+}
+
+static int mmc_bus_test(struct sd_card *card)
+{
+	/*Currently WIP*/
+	return 0;
+
+	int ret;
+	struct sdhc_command cmd = {0};
+	struct sdhc_data data = {0};
+	uint8_t* bus_test_data;
+	uint8_t* bus_test_mask;
+	uint8_t bus_8bit_test_data[8] = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xAA, 0x55};
+	uint8_t bus_4bit_test_data[4] = {0x00, 0x00, 0x00, 0x5A};
+	uint8_t bus_1bit_test_data[1] = {0x80};
+	uint8_t bus_8bit_test_mask[8] = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xFF, 0xFF};
+	uint8_t bus_4bit_test_mask[4] = {0x00, 0x00, 0x00, 0xFF};
+	uint8_t bus_1bit_test_mask[1] = {0xC0};
 
-	ret = 0;
+	switch (card->bus_io.bus_width) {
+		case SDHC_BUS_WIDTH8BIT:
+			data.blocks = 8;
+			bus_test_data = bus_8bit_test_data;
+			bus_test_mask = bus_8bit_test_mask;
+			break;
+		case SDHC_BUS_WIDTH4BIT:
```

Comment:
It seems you have support for testing 4 bit bus width, but not for setting it. Is that intentional, or just temporary?

Reply:
temporary